BlockMatrix lmmreg

jbloom22 opened the pull request:
- added LinearMixedRegresionDist and related
- added Dist tests in testSmall and testLDandRRMLOCO
- added Dist version of Eigen
- split apart full rank and low rank RDD transformations
- pushed blocked math down further in FullRank
- removed blocking in low rank
- pulled up reused matrices in low rank
- factored out globalFit function

johnc1231 commented on src/main/scala/is/hail/methods/KinshipMatrix.scala:
```diff
@@ -38,18 +39,18 @@ case class KinshipMatrix(hc: HailContext, sampleSignature: Type, matrix: Indexed
   
   def eigen(optNEigs: Option[Int]): Eigendecomposition = {
-    val K = matrix.toLocalMatrix().asBreeze().toDenseMatrix
+    val K = matrix.toLocalMatrix().asBreeze().asInstanceOf[DenseMatrix[Double]]
```
Why change this? Not a big deal since it's equivalent in this case, but toDenseMatrix on a DenseMatrix just casts to a DenseMatrix anyway.

jbloom22 replied:
asBreeze returns Matrix, and toDenseMatrix on Matrix doesn't just cast:
```
  def toDenseMatrix(implicit cm: ClassTag[V], zero: Zero[V]) = {
    DenseMatrix.tabulate(rows, cols){ (i,j) => apply(i, j)}
  }
```

jbloom22 replied:
I could change asBreeze to additionally cast to Dense or Sparse, so this wouldn't be needed in other places. Thoughts?

johnc1231 commented on src/main/scala/is/hail/methods/KinshipMatrix.scala:
```diff
@@ -66,11 +67,37 @@ case class KinshipMatrix(hc: HailContext, sampleSignature: Type, matrix: Indexed
+    val K = matrix.toLocalMatrix().asBreeze().asInstanceOf[DenseMatrix[Double]]
+
+    info(s"Computing eigenvectors of kinship matrix...")
+    val eigK = printTime(eigSymD(K))
```
We should remember to take these print times out before merging for real. I only included it in mine for the sake of experimenting.

jbloom22 replied:
Yes, but will leave for now as we need to do scaling tests

johnc1231 commented on src/main/scala/is/hail/stats/Eigendecomposition.scala:
```diff
@@ -34,4 +37,14 @@ case class Eigendecomposition(rowSignature: Type, rowIds: Array[Annotation], eve
+}
+
+case class EigendecompositionDist(rowSignature: Type, rowIds: Array[Annotation], evects: BlockMatrix, evals: DenseVector[Double]) {
+  require(evects.numRows() == rowIds.length) // FIXME: expensive
```
Not actually expensive. When you created the distributed U from local U, you used `BlockMatrixIsDistributedMatrix.from`. That constructs a BlockMatrix while specifying number of rows and columns, so accessing numRows() here is actually instantaneous.

jbloom22 replied:
Yes, and it looks like same is true after BlockMatrix mult, so I've removed the FIXME

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegression.scala:
```diff
@@ -77,279 +77,179 @@ object LinearMixedRegression {
+    val Eigendecomposition(_, rowIds, evects, evals) = eigen.filterRows(vds.sSignature, completeSamplesSet)
+
+    if (!completeSamples.sameElements(rowIds))
       fatal("Bad stuff")
```
We should be more specific here so that if a bug causes this to happen we get a message about what actually went wrong.

jbloom22 replied:
good catch. of course I didn't mean to keep "bad stuff", but I overlooked it!

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegression.scala:
```diff
@@ -60,13 +59,14 @@ object LinearMixedRegression {
     val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+    val C = cov
     val completeSamplesSet = completeSamples.toSet
     val sampleMask = vds.sampleIds.map(completeSamplesSet).toArray
```
We only technically need sampleMask if runAssoc is true. Should we move it down into the conditional for that?

johnc1231 replied:
Actually, this is true for completeSamplesSet  and completeSampleIndex as well.

jbloom22 replied:
completeSampleSet is needed for the local case when subsetting the eigendecomposition. But I moved the others down.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegression.scala:
```diff
@@ -541,23 +445,77 @@ case class DiagLMM(
-            S: DenseVector[Double], Ut: DenseMatrix[Double]): LMMConstants = {
-    val UtC = Ut * C
-    val Uty = Ut * y
+trait ScalarLMM
```
Doesn't seem like there is any point in having the trait anymore. FullRankScalarLMM and LowRankScalarLmm no longer share an interface.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegression.scala:
```diff
@@ -541,23 +445,77 @@ case class DiagLMM(
+
+    val CdC = invDelta * CtC + CzC
+
+    val b = CdC \ Cdy
```
Think you need a try catch here or you run risk of `MatrixSingularException`

johnc1231 commented on src/main/scala/is/hail/stats/ComputeRRM.scala:
```diff
@@ -83,6 +83,21 @@ object ToNormalizedRowMatrix {
+    val variantIdxBc = vds.sparkContext.broadcast(variants.index)
+    val indexedRows = 
+      if (!useDosage)
+        vds.rdd.map { case (v, (va, gs)) => IndexedRow(variantIdxBc.value(v), Vectors.dense(RegressionUtils.hardCalls(gs, n, sampleMask).toArray)) } // FIXME: handle sparse?
```
Think we can manage this without broadcasting the variant index. Can't we just `zipWithIndex` and work from there? The variants are already in order as per `OrderedRDD` backing the VDS (at least I'm pretty sure that's always the case). Just seems unnecessary to broadcast what could be kind of a big thing if there are a ton of variants.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+    val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+    val C = cov
+    val completeSamplesSet = completeSamples.toSet
+    val sampleMask = vds.sampleIds.map(completeSamplesSet).toArray
```
As I said on regular LMMReg, seems this stuff is only necessary if runAssoc is true. Maybe move it down?

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+    val EigendecompositionDist(_, rowIds, evects, evals) = eigenDist
+
+    if (!completeSamples.sameElements(rowIds))
+      fatal("Bad stuff")
```
More descriptive error needed.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+    if (!completeSamples.sameElements(rowIds))
+      fatal("Bad stuff")
+
+    val Ut = evects.transpose
```
Confirm this with Dan, but I think that that just as using `multiply` is worse than using `*` on BlockMatrices now,  I think using `transpose` is worse than using `t` as imported from `DistributedMatrix`.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+  }
+}
+
+object Helper {
```
If you are going to keep this named Helper, it cannot have global scope the way it does here. It should probably be inside the `LinearMixedRegressionDist` object anyway, regardless of name.

jbloom22 replied:
for now i've moved the methods inside LinearMixedRegressionDist

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+
+          val scalarLMMBc = sc.broadcast(scalarLMM)
+
+          val projG = (allG.toBlockMatrixDense() * (Ut :* diagLMM.sqrtInvD.toArray).transpose)
```
Don't use transpose, see above.

johnc1231 commented on src/main/scala/is/hail/methods/LinearMixedRegressionDist.scala:
```diff
@@ -0,0 +1,216 @@
+          val scalarLMM = LowRankScalarLMM(lmmConstants, diagLMM.delta, diagLMM.logNullS2, useML)
+          val scalarLMMBc = sc.broadcast(scalarLMM)
+
+          val projG = (allG.toBlockMatrixDense() * Ut.transpose)
```
Don't use transpose, see above.

jbloom22 commented:
I've addressed your comments, incorporated IntOrderedPartitioner, moved LinearMixedRegressionDist into LinearMixedRegression, done some renaming, and built more Python interfaces so I can start testing, etc
